src/WebJobs.Extensions.DurableTask/DurableOrchestrationContext.cs

```diff
@@ -254,28 +254,28 @@ public override Task<T> WaitForExternalEvent<T>(string name)
 
             lock (this.pendingExternalEvents)
             {
-                object tcsRefList;
-                List<TaskCompletionSource<T>> tcsList;
+                ICollection tcsRefCollection;
+                Queue<TaskCompletionSource<T>> tcsQueue;
                 TaskCompletionSource<T> tcs;
-                if (!this.pendingExternalEvents.TryGetValue(name, out tcsRefList) ||
-                    (tcsList = tcsRefList as List<TaskCompletionSource<T>>) == null ||
-                    (tcsList.Count == 0) ||
-                    (tcs = tcsList[0] as TaskCompletionSource<T>) == null)
+
+                if (!this.pendingExternalEvents.TryGetValue(name, out tcsRefCollection))
                 {
                     tcs = new TaskCompletionSource<T>();
-                    tcsList = new List<TaskCompletionSource<T>>
-                    {
-                        tcs,
-                    };
-                    this.pendingExternalEvents[name] = tcsList;
+                    tcsQueue = new Queue<TaskCompletionSource<T>>();
+                    tcsQueue.Enqueue(tcs);
+                    this.pendingExternalEvents[name] = tcsQueue;
                 }
                 else
                 {
-                    if (tcsList != null && tcsList.Count > 0)
+                    if ((tcsQueue = tcsRefCollection as Queue<TaskCompletionSource<T>>) == null)

```

Comment:
If we use `Queue` instead of `Queue<TaskCompletionSource<T>>`, then I think we will need to do a peek of the first element to learn whether `T` is the correct type.

Reply:
Yes, this will be great optimization. Thank you!